### ISSUE TITLE
Refactoring, testing, documentation campaign 2020/2021

### ISSUE THREAD

**Issue by edwardhartnett** [NOAA-EMC/NCEPLIBS#143]
We are engaged in a refactoring campaign with the goals of:
* Reducing overall complexity by merging related projects and deprecating code where possible.
* Adding unit testing for all operational code.
* Adding consistent and informative documentation using doxygen.

This issue can be used to keep track of our overall progress. It can be closed in about a year, when this campaign is complete.

Here's an email summary of the libraries that Mark sent me before he left:

> Some history. Back in 2012, I did a full inventory of nceplibs, determining which libraries and routines were used, and eliminated all the ones that were no longer needed. At the same time, I broke the old w3lib into w3emc and w3nco, depending on which center was responsible for each routine in order to bypass conflicts in production implementations. That worked out fine, but since then, the nco people who were responsible for their routines moved to emc so after that we could have put them back together but didn't. Anyway, that was when WCOSS1 was delivered, and now WCOSS2 is being delivered. So now is the time to again do a full inventory and reorganization of nceplibs.
> 
> w3 (now w3emc and w3nco) is the oldest set of routines, some of them coded in the 1960s even. The library is a junkbox of miscellany plus it has grib1 encoding. We would like to be rid of grib1 but there are still some customers dependent on it. No testing.
> 
> g2, g2c, and g2template are grib2 encoding and decoding libraries. They are active and important. There may be testing, Boi Vuong would know.
> 
> ip is the interpolation library for grib1 described grids, and ip2 is the interpolation library for grib2 described grids. ip2 is active and important, a modern library with testing supported by George Gayno.
> 
> sp is the spectral interpolation library, not as important after the global spectral model is gone. It has some testing. Kyle is in charge.
> 
> There are several IO libraries some of which are more legacy than others but all may go away if we go full netcdf. What will stay is bacio which is the binary stream IO needed for grib2. Supported by Hang and George V.
> 
> bufr is the observations encoding library supported by Jeff Ator. Active and important.
> 
> wgrib2 is a utility and a library and a python interface. Development is mainly done by Wesley in CPC. We are still trying to figure out the emc support.
>

**Comment by edwardhartnett** [comment]
Here's an assessment of the state of NCEPLIBS in September 2020:

![image](https://user-images.githubusercontent.com/38856240/96593287-90508800-12a6-11eb-87c1-3f83dc51b036.png)

**Comment by edwardhartnett** [comment]
Here's an assessment based on the current state:

![image](https://user-images.githubusercontent.com/38856240/96593394-b0804700-12a6-11eb-977c-0290290b3756.png)

**Comment by edwardhartnett** [comment]
Here's a chart of NCEPLIBS interdependencies. It's complex...

![image](https://user-images.githubusercontent.com/38856240/97347433-696bf600-1852-11eb-8c5b-d435abf8fa8c.png)

**Comment by edwardhartnett** [comment]
@Hang-Lei-NOAA and @BoiVuong-NOAA here's a disturbing observation. w3emc depends on nemsio, and nemsio depend on w3nco.

So when the two w3 repo are combined, it seems like we will have a circular dependency. Or am I missing something?

**Comment by BoiVuong-NOAA** [comment]
hi Ed, 
I do not know much about nemsio (Hang might).  Now, w3emc also depends on netcdf and hdf5.

**Comment by edwardhartnett** [comment]
We made a lot of progress on refactoring, testing, and documentation. I will close this issue and work will continue under other issues.
